src/React/Basic/Hooks/Suspense/Store.purs

```diff
@@ -1,41 +1,55 @@
 module React.Basic.Hooks.Suspense.Store
-  ( mkSuspenseStore
-  , SuspenseStore
+  ( SuspenseStore
+  , mkSuspenseStore
   , get
   , get'
+  , class HasBackend
+  , fromKey
+  , backend
   ) where
 
 import Prelude
+
 import Control.Alt ((<|>))
-import Data.DateTime.Instant (Instant, unInstant)
+import Data.DateTime.Instant (unInstant)
 import Data.Either (Either(..))
+import Data.Function (on)
 import Data.Int (ceil)
-import Data.Map (Map)
 import Data.Map as Map
 import Data.Maybe (Maybe(..))
 import Data.Newtype (un)
+import Data.Ord (greaterThan)
+import Data.String (joinWith)
+import Data.Symbol (class IsSymbol, SProxy(..), reflectSymbol)
 import Data.Time.Duration (Milliseconds(..))
+import Data.Tuple (fst, snd)
 import Effect (Effect)
 import Effect.Aff (Aff, attempt, launchAff, throwError)
 import Effect.Class (liftEffect)
 import Effect.Console (warn)
 import Effect.Exception (try)
 import Effect.Now (now)
-import Effect.Ref (Ref)
 import Effect.Ref as Ref
-import React.Basic.Hooks (type (/\), (/\))
+import React.Basic.Hooks ((/\))
 import React.Basic.Hooks.Suspense (Suspended(..), SuspenseResult(..))
+import Unsafe.Coerce (unsafeCoerce)
 import Web.HTML (window)
 import Web.HTML.Window (requestIdleCallback)
 
--- | Simple key-based cache.
+get :: forall k v s. HasBackend k v s => SuspenseStore -> k -> Suspended v
+get s k = _get s Nothing k
+
+get' :: forall k v s. HasBackend k v s => SuspenseStore -> Milliseconds -> k -> Suspended v
+get' s d k = _get s (Just d) k
+
+class IsSymbol s <= HasBackend k v (s :: Symbol) | k -> s, k -> v where
+  fromKey :: k -> String
+  backend :: k -> Aff v
```

Comment:
Implementing `HasBackend` and calling the getters above is the external API. The classes and types below should contain the unsafe bits.

Reply:
I'm not sure this will do what you want it to: with this code, I think I can define instances
```
HasBackend String Int "1"
HasBackend Int Int "1"
```

since we are effectively only saying that there can only be one HasBackend instance for each key type `k`; we aren't preventing you from defining two different instances with different `k` types but the same `s` type. A functional dependency `s -> k` would prevent you from doing this, but that probably isn't workable because you'll fall foul of orphan instance restrictions there - since in this case `s` determines everything, then the instance can only be defined in the same module as the HasBackend class. Normally you'd also be able to define the instance in the same module as the type `s`, but in this case `s` is built in to the compiler so you can't put it there. See https://liamgoodacre.github.io/purescript/type/class/instance/orphan/functional/dependencies/2017/01/22/purescript-orphan-instance-detection.html for more information on how functional dependencies affect orphan instance checking. 

I think you'd ideally need something like Haskell's Data.Typeable to do this safely, but PureScript doesn't have that. I'm not sure there's a good way of doing this without Data.Typeable, unfortunately.